### ISSUE TITLE
Bert baseline variation in result

### ISSUE THREAD

**Issue by rusiaaman** [google-research/uda#6]
For IMDB semi-supervised case:
* Given that baseline gets trained on (unaugmented) 20 examples, it doesn't make sense to me to train it for 3000 (+300 warm up) steps with 32 batch size (looking at run_base.sh). Am I missing something? 
* I ran evaluation every 3 steps (+3 warm up), and found following mildly surprising result. (Bert base, max_seq_length=512, train_batch_size=6)
  - 3 steps (~1 epoch):
    train_loss = 0.77
    eval_classify_loss = 0.69
    eval_classify_accuracy = 0.56
  - 6 steps (~2 epoch):
    train_loss = 0.61
    eval_classify_loss = 0.82
    eval_classify_accuracy = 0.56
  - 9 steps (~3 epoch):
    train_loss = 0.50
    eval_classify_loss = 0.74
    eval_classify_accuracy = 0.55
  - 12 steps (~4 epoch):
    train_loss = 0.44
    eval_classify_loss = 0.65
    eval_classify_accuracy = 0.59
  - 15 steps (~5 epoch):
    train_loss = 0.18
    eval_classify_loss = 0.48
    eval_classify_accuracy = 0.78
  - 18 steps (~6 epoch):
    train_loss = 0.08
    eval_classify_loss = 0.61
    eval_classify_accuracy = 0.75
  - 21 steps (~6 epoch):
    train_loss = 0.02
    eval_classify_loss = 0.79
    eval_classify_accuracy = 0.77
.
.
.

  - 21 steps (~6 epoch):
    train_loss = 0.0009
    eval_classify_loss = 0.93
    eval_classify_accuracy = 0.82

---
Although the best accuracy is 82%, it has overfitted. Before overfitting the best accuracy is 78% (15 steps). This is more than the reported accuracy of 72.56% (Table 1), but I suspect there is a lot of variation. How were these results reported, did you take mean/median of a few experiments? How many steps was the BERT base trained for?

Colab notebook with the output:
https://colab.research.google.com/drive/1ysEZeLFmZDUy39ip1J1VLDKLhyUNMss3

**Comment by qizhex** [comment]
We use a batch_size of 32 to be consistent with our experiments on TPU v3-32 Pod (The smallest batch_size on it is 32). The num_train_steps is set to 3000 for our experiments. And yes. There is a lot of variation. We run 10 experiments and take the max. 

The improvement of UDA should be very consistent despite the variance. Have you tried UDA with a similar setup? You might need to train more steps.

**Comment by rusiaaman** [comment]
Sorry for the confusion. I have run only the baseline (without UDA) and the above issue pertains to that. Specifically I don't see why the train_steps is 3000 for *baseline* model which is purely supervised and trained on 20 train examples. I suppose early stopping is used?

**Comment by qizhex** [comment]
Yes, early stopping is used. We evaluate the model 20 times throughout training. 

One reason of training for more steps is that we use batch_size=32, so it requires more steps. The other reason is that for experiments with BERT initialization, we adopted hyperparameters of BERT finetuned on in-domain unlabeled data and training longer yielded better performance in single experiments for finetuned BERT when we randomly sampled training steps.

**Comment by rusiaaman** [comment]
Thanks for the clarifications.
